File: compiler/packages/babel-plugin-react-compiler/src/Validation/ValidateLocalsNotReassignedAfterRender.ts
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {CompilerError, Effect} from '..';
import {HIRFunction, IdentifierId, Place} from '../HIR';
import {
  eachInstructionLValue,
  eachInstructionValueOperand,
  eachTerminalOperand,
} from '../HIR/visitors';
import {getFunctionCallSignature} from '../Inference/InferReferenceEffects';

/**
 * Validates that local variables cannot be reassigned after render.
 * This prevents a category of bugs in which a closure captures a
 * binding from one render but does not update
 */
export function validateLocalsNotReassignedAfterRender(fn: HIRFunction): void {
  const contextVariables = new Set<IdentifierId>();
  const reassignment = getContextReassignment(
    fn,
    contextVariables,
    false,
    false,
  );
  if (reassignment !== null) {
    CompilerError.throwInvalidReact({
      reason:
        'Reassigning a variable after render has completed can cause inconsistent behavior on subsequent renders. Consider using state instead',
      description:
        reassignment.identifier.name !== null &&
        reassignment.identifier.name.kind === 'named'
          ? `Variable \`${reassignment.identifier.name.value}\` cannot be reassigned after render`
          : '',
      loc: reassignment.loc,
    });
  }
}

function getContextReassignment(
  fn: HIRFunction,
  contextVariables: Set<IdentifierId>,
  isFunctionExpression: boolean,
  isAsync: boolean,
): Place | null {
  const reassigningFunctions = new Map<IdentifierId, Place>();
  for (const [, block] of fn.body.blocks) {
    for (const instr of block.instructions) {
      const {lvalue, value} = instr;
      switch (value.kind) {
        case 'FunctionExpression':
        case 'ObjectMethod': {
          let reassignment = getContextReassignment(
            value.loweredFunc.func,
            contextVariables,
            true,
            isAsync || value.loweredFunc.func.async,
          );
          if (reassignment === null) {
            // If the function itself doesn't reassign, does one of its dependencies?
            for (const operand of eachInstructionValueOperand(value)) {
              const reassignmentFromOperand = reassigningFunctions.get(
                operand.identifier.id,
              );
              if (reassignmentFromOperand !== undefined) {
                reassignment = reassignmentFromOperand;
                break;
              }
            }
          }
          // if the function or its depends reassign, propagate that fact on the lvalue
          if (reassignment !== null) {
            if (isAsync || value.loweredFunc.func.async) {
              CompilerError.throwInvalidReact({
                reason:
                  'Reassigning a variable in an async function can cause inconsistent behavior on subsequent renders. Consider using state instead',
                description:
                  reassignment.identifier.name !== null &&
                  reassignment.identifier.name.kind === 'named'
                    ? `Variable \`${reassignment.identifier.name.value}\` cannot be reassigned after render`
                    : '',
                loc: reassignment.loc,
              });
            }
            reassigningFunctions.set(lvalue.identifier.id, reassignment);
          }
          break;
        }
        case 'StoreLocal': {
          const reassignment = reassigningFunctions.get(
            value.value.identifier.id,
          );
          if (reassignment !== undefined) {
            reassigningFunctions.set(
              value.lvalue.place.identifier.id,
              reassignment,
            );
            reassigningFunctions.set(lvalue.identifier.id, reassignment);
          }
          break;
        }
        case 'LoadLocal': {
          const reassignment = reassigningFunctions.get(
            value.place.identifier.id,
          );
          if (reassignment !== undefined) {
            reassigningFunctions.set(lvalue.identifier.id, reassignment);
          }
          break;
        }
        case 'DeclareContext': {
          if (!isFunctionExpression) {
            contextVariables.add(value.lvalue.place.identifier.id);
          }
          break;
        }
        case 'StoreContext': {
          if (isFunctionExpression) {
            if (contextVariables.has(value.lvalue.place.identifier.id)) {
              return value.lvalue.place;
            }
          } else {
            /*
             * We only track reassignments of variables defined in the outer
             * component or hook.
             */
            contextVariables.add(value.lvalue.place.identifier.id);
          }
          const reassignment = reassigningFunctions.get(
            value.value.identifier.id,
          );
          if (reassignment !== undefined) {
            reassigningFunctions.set(
              value.lvalue.place.identifier.id,
              reassignment,
            );
            reassigningFunctions.set(lvalue.identifier.id, reassignment);
          }
          break;
        }
        default: {
          let operands = eachInstructionValueOperand(value);
          // If we're calling a function that doesn't let its arguments escape, only test the callee
          if (value.kind === 'CallExpression') {
            const signature = getFunctionCallSignature(
              fn.env,
              value.callee.identifier.type,
            );
            if (signature?.noAlias) {
              operands = [value.callee];
            }
          } else if (value.kind === 'MethodCall') {
            const signature = getFunctionCallSignature(
              fn.env,
              value.property.identifier.type,
            );
            if (signature?.noAlias) {
              operands = [value.receiver, value.property];
            }
          }
          for (const operand of operands) {
            CompilerError.invariant(operand.effect !== Effect.Unknown, {
              reason: `Expected effects to be inferred prior to ValidateLocalsNotReassignedAfterRender`,
              loc: operand.loc,
            });
            const reassignment = reassigningFunctions.get(
              operand.identifier.id,
            );
            if (reassignment !== undefined) {
              /*
               * Functions that reassign local variables are inherently mutable and are unsafe to pass
               * to a place that expects a frozen value. Propagate the reassignment upward.
               */
              if (operand.effect === Effect.Freeze) {
                return reassignment;
              } else {
                /*
                 * If the operand is not frozen but it does reassign, then the lvalues
                 * of the instruction could also be reassigning
                 */
                for (const lval of eachInstructionLValue(instr)) {
                  reassigningFunctions.set(lval.identifier.id, reassignment);
                }
              }
            }
          }
          break;
        }
      }
    }
    for (const operand of eachTerminalOperand(block.terminal)) {
      const reassignment = reassigningFunctions.get(operand.identifier.id);
      if (reassignment !== undefined) {
        return reassignment;
      }
    }
  }
  return null;
}

File: compiler/crates/react_compiler_validation/src/validate_locals_not_reassigned_after_render.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::{HashMap, HashSet};

use react_compiler_diagnostics::{
    CompilerDiagnostic, CompilerDiagnosticDetail, ErrorCategory,
};
use react_compiler_hir::environment::Environment;
use react_compiler_hir::{
    Effect, HirFunction, Identifier, IdentifierId, IdentifierName, InstructionValue,
    Place, Type,
};
use react_compiler_hir::visitors::{each_instruction_lvalue, each_instruction_value_operand, each_terminal_operand};

/// Validates that local variables cannot be reassigned after render.
/// This prevents a category of bugs in which a closure captures a
/// binding from one render but does not update.
pub fn validate_locals_not_reassigned_after_render(func: &HirFunction, env: &mut Environment) {
    let mut context_variables: HashSet<IdentifierId> = HashSet::new();
    let mut diagnostics: Vec<CompilerDiagnostic> = Vec::new();

    let reassignment = get_context_reassignment(
        func,
        &env.identifiers,
        &env.types,
        &env.functions,
        env,
        &mut context_variables,
        false,
        false,
        &mut diagnostics,
    );

    // Record accumulated errors (from async function checks in inner functions) first
    for diagnostic in diagnostics {
        env.record_diagnostic(diagnostic);
    }

    // Then record the top-level reassignment error if any
    if let Some(reassignment_place) = reassignment {
        let variable_name = format_variable_name(&reassignment_place, &env.identifiers);
        env.record_diagnostic(
            CompilerDiagnostic::new(
                ErrorCategory::Immutability,
                "Cannot reassign variable after render completes",
                Some(format!(
                    "Reassigning {} after render has completed can cause inconsistent \
                     behavior on subsequent renders. Consider using state instead",
                    variable_name
                )),
            )
            .with_detail(CompilerDiagnosticDetail::Error {
                loc: reassignment_place.loc,
                message: Some(format!(
                    "Cannot reassign {} after render completes",
                    variable_name
                )),
            }),
        );
    }
}

/// Format a variable name for error messages. Uses the named identifier if
/// available, otherwise falls back to "variable".
fn format_variable_name(place: &Place, identifiers: &[Identifier]) -> String {
    let identifier = &identifiers[place.identifier.0 as usize];
    match &identifier.name {
        Some(IdentifierName::Named(name)) => format!("`{}`", name),
        _ => "variable".to_string(),
    }
}

/// Check whether a function type has a noAlias signature.
fn has_no_alias_signature(
    env: &Environment,
    identifier_id: IdentifierId,
    identifiers: &[Identifier],
    types: &[Type],
) -> bool {
    let ty = &types[identifiers[identifier_id.0 as usize].type_.0 as usize];
    env.get_function_signature(ty)
        .ok()
        .flatten()
        .map_or(false, |sig| sig.no_alias)
}

/// Recursively checks whether a function (or its dependencies) reassigns a
/// context variable. Returns the reassigned place if found, or None.
///
/// Side effects: accumulates async-function reassignment diagnostics into `diagnostics`.
fn get_context_reassignment(
    func: &HirFunction,
    identifiers: &[Identifier],
    types: &[Type],
    functions: &[HirFunction],
    env: &Environment,
    context_variables: &mut HashSet<IdentifierId>,
    is_function_expression: bool,
    is_async: bool,
    diagnostics: &mut Vec<CompilerDiagnostic>,
) -> Option<Place> {
    // Maps identifiers to the place that they reassign
    let mut reassigning_functions: HashMap<IdentifierId, Place> = HashMap::new();

    for (_block_id, block) in &func.body.blocks {
        for &instruction_id in &block.instructions {
            let instr = &func.instructions[instruction_id.0 as usize];

            match &instr.value {
                InstructionValue::FunctionExpression { lowered_func, .. }
                | InstructionValue::ObjectMethod { lowered_func, .. } => {
                    let inner_function = &functions[lowered_func.func.0 as usize];
                    let inner_is_async = is_async || inner_function.is_async;

                    // Recursively check the inner function
                    let mut reassignment = get_context_reassignment(
                        inner_function,
                        identifiers,
                        types,
                        functions,
                        env,
                        context_variables,
                        true,
                        inner_is_async,
                        diagnostics,
                    );

                    // If the function itself doesn't reassign, check if one of its
                    // dependencies (operands) is a reassigning function
                    if reassignment.is_none() {
                        for context_place in &inner_function.context {
                            if let Some(reassignment_place) =
                                reassigning_functions.get(&context_place.identifier)
                            {
                                reassignment = Some(reassignment_place.clone());
                                break;
                            }
                        }
                    }

                    // If the function or its dependencies reassign, handle it
                    if let Some(ref reassignment_place) = reassignment {
                        if inner_is_async {
                            // Async functions that reassign get an immediate error
                            let variable_name =
                                format_variable_name(reassignment_place, identifiers);
                            diagnostics.push(
                                CompilerDiagnostic::new(
                                    ErrorCategory::Immutability,
                                    "Cannot reassign variable in async function",
                                    Some(
                                        "Reassigning a variable in an async function can cause \
                                         inconsistent behavior on subsequent renders. \
                                         Consider using state instead"
                                            .to_string(),
                                    ),
                                )
                                .with_detail(CompilerDiagnosticDetail::Error {
                                    loc: reassignment_place.loc,
                                    message: Some(format!(
                                        "Cannot reassign {}",
                                        variable_name
                                    )),
                                }),
                            );
                            // Return null (don't propagate further) — matches TS behavior
                        } else {
                            // Propagate reassignment info on the lvalue
                            reassigning_functions
                                .insert(instr.lvalue.identifier, reassignment_place.clone());
                        }
                    }
                }

                InstructionValue::StoreLocal { lvalue, value, .. } => {
                    if let Some(reassignment_place) =
                        reassigning_functions.get(&value.identifier)
                    {
                        let reassignment_place = reassignment_place.clone();
                        reassigning_functions
                            .insert(lvalue.place.identifier, reassignment_place.clone());
                        reassigning_functions
                            .insert(instr.lvalue.identifier, reassignment_place);
                    }
                }

                InstructionValue::LoadLocal { place, .. }
                | InstructionValue::LoadContext { place, .. } => {
                    if let Some(reassignment_place) =
                        reassigning_functions.get(&place.identifier)
                    {
                        reassigning_functions
                            .insert(instr.lvalue.identifier, reassignment_place.clone());
                    }
                }

                InstructionValue::DeclareContext { lvalue, .. } => {
                    if !is_function_expression {
                        context_variables.insert(lvalue.place.identifier);
                    }
                }

                InstructionValue::StoreContext { lvalue, value, .. } => {
                    // If we're inside a function expression and the target is a
                    // context variable from the outer scope, this is a reassignment
                    if is_function_expression
                        && context_variables.contains(&lvalue.place.identifier)
                    {
                        return Some(lvalue.place.clone());
                    }

                    // In the outer function, track context variables
                    if !is_function_expression {
                        context_variables.insert(lvalue.place.identifier);
                    }

                    // Propagate reassigning function info through StoreContext
                    if let Some(reassignment_place) =
                        reassigning_functions.get(&value.identifier)
                    {
                        let reassignment_place = reassignment_place.clone();
                        reassigning_functions
                            .insert(lvalue.place.identifier, reassignment_place.clone());
                        reassigning_functions
                            .insert(instr.lvalue.identifier, reassignment_place);
                    }
                }

                _ => {
                    // For calls with noAlias signatures, only check the callee/receiver
                    // (not args) to avoid false positives from callbacks that reassign
                    // context variables.
                    let operands: Vec<Place> = match &instr.value {
                        InstructionValue::CallExpression { callee, .. } => {
                            if has_no_alias_signature(
                                env,
                                callee.identifier,
                                identifiers,
                                types,
                            ) {
                                vec![callee.clone()]
                            } else {
                                each_instruction_value_operand(&instr.value, env)
                            }
                        }
                        InstructionValue::MethodCall {
                            receiver, property, ..
                        } => {
                            if has_no_alias_signature(
                                env,
                                property.identifier,
                                identifiers,
                                types,
                            ) {
                                vec![receiver.clone(), property.clone()]
                            } else {
                                each_instruction_value_operand(&instr.value, env)
                            }
                        }
                        InstructionValue::TaggedTemplateExpression { tag, .. } => {
                            if has_no_alias_signature(
                                env,
                                tag.identifier,
                                identifiers,
                                types,
                            ) {
                                vec![tag.clone()]
                            } else {
                                each_instruction_value_operand(&instr.value, env)
                            }
                        }
                        _ => each_instruction_value_operand(&instr.value, env),
                    };

                    for operand in &operands {
                        // Invariant: effects must be inferred before this pass runs
                        assert!(
                            operand.effect != Effect::Unknown,
                            "Expected effects to be inferred prior to \
                             ValidateLocalsNotReassignedAfterRender"
                        );

                        if let Some(reassignment_place) =
                            reassigning_functions.get(&operand.identifier).cloned()
                        {
                            if operand.effect == Effect::Freeze {
                                // Functions that reassign local variables are inherently
                                // mutable and unsafe to pass where a frozen value is expected.
                                return Some(reassignment_place);
                            } else {
                                // If the operand is not frozen but does reassign, then the
                                // lvalues of the instruction could also be reassigning
                                for lvalue_id in each_instruction_lvalue_ids(instr) {
                                    reassigning_functions
                                        .insert(lvalue_id, reassignment_place.clone());
                                }
                            }
                        }
                    }
                }
            }
        }

        // Check terminal operands for reassigning functions
        for operand in each_terminal_operand(&block.terminal) {
            if let Some(reassignment_place) = reassigning_functions.get(&operand.identifier) {
                return Some(reassignment_place.clone());
            }
        }
    }

    None
}

/// Collect all lvalue identifier IDs from an instruction.
/// Thin wrapper around canonical `each_instruction_lvalue` that maps to ids.
fn each_instruction_lvalue_ids(instr: &react_compiler_hir::Instruction) -> Vec<IdentifierId> {
    each_instruction_lvalue(instr)
        .into_iter()
        .map(|p| p.identifier)
        .collect()
}
